###**Student & UHI Submission Details**

**Name:** Alanna Zimbehl  
**Student ID:** 19016467  
**Module:** Data Analytics on the Web  
**Assignment:** Assignment Two - Data Science (Python)

---

#Assignment Two - Introduction
Continuing on from the first section of the assignment, this notebook will contain the process and analysis of training various models, as well as analysing the accuracy of the predictions those models are making. 

#Linear Regression and Neural Networks

**Neural Networks**

A neural network is a deep learning technique which are comprised of multiple algorithms to detec patterns in data without having to actually manually coding an entire network. They're main idea was to mimick the process of a brain where each node acts a neuron, connected to many other neurons firing signals in this case passing data like a network, learning and repeating the process until it's completed. (IBM, 2023)

**Linear Regressions**  
Linear Regression is another model that is used in machine learning, where the model will try predict the value of a variable based on the value of another variable. This is ideal to find relationships between variables and can otherwise be known as dependant and independant variables, where indepentant variables are used to predict the dependant variables. These particular models are ideal for making forecasting predictions or for spotting trends whether they are negative or positive. Typically there are visually depicted as scatter plots with various data points plotted and a calculation is perform to find the best line fit. This can help determine positive or negative corelatoins, these can also be weak where the data is quite scatter and has little meaning or strong where the data is relatively close together to signify importance. (IBM,2023)


#Results
##Linear Regression Models

In [ ]:
#import starting libraries
import pandas as pd
import numpy as np
import tensorflow as tf

#checking tensorflow
print(tf.__version__)


2.12.0


To begin the training of the models, first need to import three libraries which will handle much of the process. To test the libraries have imported correctly, it's best to print out their installed versions which will aid in finding the correct documentation too.

In [ ]:
#loading the full dataset.
df_full = pd.read_csv('https://raw.githubusercontent.com/Behls/data-anayltics-uhi/main/assignmentTwo/data/collated_data_cleaned_final.csv')
print(df_full)
#loading the 2014 dataset.
df_2014 = pd.read_csv('https://raw.githubusercontent.com/Behls/data-anayltics-uhi/main/assignmentTwo/data/collated_data_year_2014.csv')
print(df_2014)

      day  year  mo  da collision_date  temp  dewp     slp  visib  wdsp  \
0       4  2020  12  31     31/12/2020  57.4  56.2  1024.4    2.0   5.6   
1       3  2020  12  30     30/12/2020  73.8  65.7  1015.8    8.3  10.1   
2       2  2020  12  29     29/12/2020  65.0  63.4  1013.5    6.6  12.0   
3       1  2020  12  28     28/12/2020  68.8  60.1  1013.0    9.4   7.6   
4       7  2020  12  27     27/12/2020  63.5  61.3  1011.9    2.3   5.0   
...   ...   ...  ..  ..            ...   ...   ...     ...    ...   ...   
3101    4  2012   7   5     05/07/2012  34.2  30.8  1011.1    6.4  11.9   
3102    3  2012   7   4     04/07/2012  36.5  25.7  1007.3   10.0  17.9   
3103    2  2012   7   3     03/07/2012  42.8  38.5  1019.4   10.0   9.8   
3104    1  2012   7   2     02/07/2012  40.4  24.6  1012.4   10.0  13.9   
3105    7  2012   7   1     01/07/2012  36.3  20.4  1018.1   10.0  12.6   

      mxpsd  gust   max   min  prcp  sndp  fog  num_collisions  
0       8.9   0.0  69.1  53.1  0.0

Loading the full and seperated 2014 dataset from github, and then printing out the data to ensure all columns and rows are correct.

In [ ]:
#splitting each dataset - full dataset
inputs_full = [df_full['da'], df_full['mo'], df_full['temp'], df_full['prcp'], df_full['dewp'], df_full['num_collisions']]
headers_full = ["day", "month", "temperature", "percipitation",  "dew point", "number collisions"]
combined = pd.concat(inputs_full, axis=1, keys=headers_full)
print(combined)

#splitting each dataset - 2014 dataset
inputs_2014 = [df_2014['da'],df_2014['mo'],df_2014['temp'],df_2014['max'],df_2014['min'],df_2014['prcp'], df_2014['wdsp'], df_2014['dewp'],df_2014['slp'], df_2014['num_collisions']]
headers_2014 = ["day", "month", "temperature","max temperature", "min temperature", "percipitation", "wind speed", "dew point", "sea-level pressure", "number collisions"]
combined_2014 = pd.concat(inputs_2014, axis=1, keys=headers_2014)
print(combined_2014)



      day  month  temperature  percipitation  dew point  number collisions
0      31     12         57.4           0.00       56.2                278
1      30     12         73.8           0.00       65.7                238
2      29     12         65.0           0.05       63.4                244
3      28     12         68.8           0.00       60.1                217
4      27     12         63.5           0.00       61.3                184
...   ...    ...          ...            ...        ...                ...
3101    5      7         34.2           0.07       30.8                591
3102    4      7         36.5           0.35       25.7                432
3103    3      7         42.8           0.00       38.5                664
3104    2      7         40.4           0.22       24.6                564
3105    1      7         36.3           0.01       20.4                538

[3106 rows x 6 columns]
     day  month  temperature  max temperature  min temperature  percipitati

Then once the datasets are correctly imported, the features that will train and test the models will need to be saved into a new dataframe. This applies to both datasets. Additionally new headers would be applied to the dataset, which makes it a little easier to read where they will last be combined using pandas built in concat function.

In [ ]:
#splitting train/test from full dataset
train_full = combined.sample(frac=0.80, random_state=0)
test_full = combined.drop(train_full.index)

#splitting test/train from 2014 datset
train_2014 = combined_2014.sample(frac=0.80, random_state=0)
test_2014 = combined_2014.drop(train_2014.index)

Each dataset will be further seperated into testing and training datasets, where there should be a total of four. Typically the training data is 80% and the testing data is 20% of the whole original dataset, however due to their small sizes these percentages may changed depending on the results from the models. Models can be overfitted, where they perform well during training but poorly during testing, so it would be benifical to allow these percentages to be flexible.

In [ ]:
#print the test/train datasets
print(train_full)
print(test_full)
print(train_2014)
print(test_2014)

      day  month  temperature  percipitation  dew point  number collisions
2554    3      1         33.4           0.36       27.2                423
253    22      4         31.1           0.00       11.6                148
1651   24      6         50.4           0.00       43.6                811
791     1     11         43.6           0.00       27.9                696
570    10      6         66.5           0.00       60.0                613
...   ...    ...          ...            ...        ...                ...
337    29      1         43.2           0.00       36.1                495
46     15     11         66.2           0.00       56.9                251
747    15     12         65.6           0.00       54.0                628
2836   27      3         35.4           0.00       25.4                441
753     9     12         73.5           0.00       59.6                465

[2485 rows x 6 columns]
      day  month  temperature  percipitation  dew point  number collisions


To ensure the training and testing datasets have been correctly split, it's best to print and examine them for potential mistakes.

In [ ]:
#create copies of the datasets for training and testing - 2014 dataset
train_2014_copy = train_2014.copy()
test_2014_copy = test_2014.copy()

#create copies of the datasets for training and testing - full dataset
train_full_copy = train_full.copy()
test_full_copy = test_full.copy()

#creating labels - test
testing_labels_full = test_full_copy.pop('number collisions')
testing_labels_2014 = test_2014_copy.pop('number collisions')

#creating labels - train
training_labels_full = train_full_copy.pop('number collisions')
training_labels_2014 = train_2014_copy.pop('number collisions')

#testing labels work against dataset
print(train_2014_copy)
print(train_full_copy)
print(training_labels_2014)



     

     day  month  temperature  max temperature  min temperature  percipitation  \
106   16      9         53.6             61.0             50.0           0.05   
259   16      4         64.9             73.9             54.0           0.00   
45    16     11         71.1             77.0             68.0           0.01   
26     5     12         77.4             87.1             69.1           0.00   
78    14     10         69.4             73.9             64.9           0.01   
..   ...    ...          ...              ...              ...            ...   
227   18      5         55.2             64.0             46.0           0.00   
148    5      8         51.3             60.1             44.1           0.08   
143   10      8         59.0             66.2             53.1           0.00   
180    4      7         42.1             48.0             28.0           0.55   
131   22      8         55.4             59.0             52.0           0.10   

     wind speed  dew point 

Finally this is the stage where the features are seperated from the labels otherwise known as outputs, so the model can train on features seperately and the ouputs to learn to predict outputs based on data passed to it. In this case the model is designed to predict the number of collisions based on various features, therefor the number of collisions column needs to be removed and stored in a label variable to used during training and testing. This will essentially split the dataset further, however it's also imported to make copies of the previous datasets so the originals remain untouched incase of errors.

In [ ]:
#linear regression - full
normalisation_layer = tf.keras.layers.Normalization(input_shape=[5,], axis=None)
normalisation_layer.adapt(np.array(train_full_copy))

model_full = tf.keras.Sequential([
 normalisation_layer,
 tf.keras.layers.Dense(units=5)   
])

model_full.compile(optimizer = tf.optimizers.Adam(learning_rate=0.2),
                   loss="mean_squared_error"
                   )

model_output = model_full.fit(train_full_copy, training_labels_full, epochs=100, validation_split=0.25)


Epoch 1/100
59/59 [==============================] - 1s 5ms/step - loss: 302077.9688 - val_loss: 264238.1875
Epoch 2/100
59/59 [==============================] - 0s 3ms/step - loss: 243514.8750 - val_loss: 212465.2812
Epoch 3/100
59/59 [==============================] - 0s 3ms/step - loss: 195602.4844 - val_loss: 170238.0000
Epoch 4/100
59/59 [==============================] - 0s 3ms/step - loss: 156728.9844 - val_loss: 136659.6719
Epoch 5/100
59/59 [==============================] - 0s 3ms/step - loss: 125936.3516 - val_loss: 110193.4141
Epoch 6/100
59/59 [==============================] - 0s 3ms/step - loss: 101803.4062 - val_loss: 89687.2656
Epoch 7/100
59/59 [==============================] - 0s 3ms/step - loss: 83292.0547 - val_loss: 74297.3828
Epoch 8/100
59/59 [==============================] - 0s 3ms/step - loss: 69433.7500 - val_loss: 62910.3203
Epoch 9/100
59/59 [==============================] - 0s 3ms/step - loss: 59148.5156 - val_loss: 54569.1172
Epoch 10/100
59/59 [======

At this stage it's a case of setting up the configuration such as normalisation for the model. In this case as the full dataset contains five feature columns the normalisation process would ideally take five inputs at a time as it needs to convert five features to predict one output value. As this dataset is relatively small it wouldn't require much processing power or time to do at least 100 passes on each sample. The normalisation layer takes a range of values, finds the minimum and then the maxmimum assigns them values bet 0 and -1 and every value inbetween it calculates a normalised value between those figures. During the analysis stage there where two loss functions that where chosen to measure the errors within the dataset, first being mean squared error and the second being root mean squared error. 

The linear regression model appears to be preforming quite poorly, ideally would like to see lower loss numbers and these values are closely validated by the validation loss therefor it's likely this model's accuracy making predictions will be quite poor. In this particular model the mean squared error loss function was used, where it will find the mean squared errors between the actual real values and ones that are predicted. (Neptune AI, 2023) The epochs where increased from 10 to 100 so it makes more passes over the datset, and that didn't return any noticable improvements to the accuracy of the model. Originall the epochs where set to 1000 however it was evident the loss values didnt decrease enough for it to be viable, additionally the different between the loss value and validation value increase where it should infact decrease. Additionally the learning rate was increased which did improve the returned results slightly.

In [ ]:
#new data
new_data = np.array([30,12,57,0.00,56])
predictions = model_full.predict(new_data)
print(predictions)

1/1 [==============================] - 0s 22ms/step
[[517.1939  517.31696 517.2752  517.24243 517.46106]]


Once the model was trained a new dataset with similar values would be passed into the model to predict the number of collisions. According to the results it perform quite poorly. The dataframe consists of intial date of a day, which is the 30th, followed by the month 12 which is december, temperature and percipitation values of 56°F and 0.00, finally followed by the dew point for the day which was 56°F. It should predict a value between 278 and 244 as it is based on the similar values before and after in the dataset, however based on the results of 517 it has performed very poorly and isn't accurate.

In [ ]:
#linear regression - 2014
normalisation_layer_2014 = tf.keras.layers.Normalization(input_shape=[9,], axis=None)
normalisation_layer_2014.adapt(np.array(train_2014_copy))

model_2014 = tf.keras.Sequential([
 normalisation_layer_2014,
 tf.keras.layers.Dense(units=10)   
])

model_2014.compile(optimizer = tf.optimizers.Adam(learning_rate=0.1),
                   loss="mean_squared_error",
                   metrics = [tf.keras.metrics.RootMeanSquaredError(name='rmse')]
                   )

model_output = model_2014.fit(train_2014_copy, training_labels_2014, epochs=100, validation_split=0.2, verbose=1)

Epoch 1/100
8/8 [==============================] - 1s 27ms/step - loss: 328058.2188 - rmse: 572.7636 - val_loss: 320477.4375 - val_rmse: 566.1072
Epoch 2/100
8/8 [==============================] - 0s 9ms/step - loss: 322100.7812 - rmse: 567.5392 - val_loss: 314573.9375 - val_rmse: 560.8689
Epoch 3/100
8/8 [==============================] - 0s 10ms/step - loss: 316201.4375 - rmse: 562.3179 - val_loss: 308767.3125 - val_rmse: 555.6683
Epoch 4/100
8/8 [==============================] - 0s 21ms/step - loss: 310405.7812 - rmse: 557.1407 - val_loss: 303015.1875 - val_rmse: 550.4682
Epoch 5/100
8/8 [==============================] - 0s 17ms/step - loss: 304657.7188 - rmse: 551.9580 - val_loss: 297342.3750 - val_rmse: 545.2911
Epoch 6/100
8/8 [==============================] - 0s 18ms/step - loss: 298982.2812 - rmse: 546.7927 - val_loss: 291759.2500 - val_rmse: 540.1475
Epoch 7/100
8/8 [==============================] - 0s 10ms/step - loss: 293395.2812 - rmse: 541.6598 - val_loss: 286234.4688 

Much like the previous linear regression, this one is training on a specific dataset which year is 2014. There is far less data, therefore far less for the model to train on. This dataset has nine columns, therefor the input shape is different to reflect that, furthermore the model is utilising the metric option moot mean square error. Both loss functions penalise larger errors however the difference mainly between the two is that the root square mean error function is measured in similar form to the predicted value. According to the results, there are quite large descrepencies between loss and validation values, where it's potentially performing slightly better in the training however it would suggest it would perform poorer those values where validated. The rsme value is still too high which again suggests the model will not perform accurate and could be off the predicted value of around 183 - 193.

In [ ]:
#new data
new_data_v2 = np.array([16,12,60,64,57,0.1,17,54,1027])
predictions = model_2014.predict(new_data_v2)
print(predictions)



1/1 [==============================] - 0s 51ms/step
[[393.49387 394.6226  394.9938  394.7438  395.607   395.08545 395.0645
  393.7133  394.00388 394.24844]]


For this particular prediction test it takes in more values that the previous test. This one includes more features such as sea level pressure, wind speed, max and min temperatures. Based on similar data from the actual datasets and values, the model should predict the number of collisions between 555 and 720 which is quite a large range, however it is under performing based on the values passed in. Therefor like before despite changing the amount of data to a more specific time period and adding additional features to the model, it's still not accurate and performing rather poor.

##Deep Neural Networks

In [76]:
#SCALE
SCALE = 1

#scaling number of collisions - 2014
training_labels_2014 = training_labels_2014/SCALE
testing_labels_2014 = testing_labels_2014/SCALE

#scaling number of collisions - full
training_labels_full = training_labels_full/SCALE
testing_labels_full = testing_labels_full/SCALE

To better normalise some of the labels, it can be better to determine a scale and divide it by the scale value, to give the model a better chance of predicting accurately. In this case, each set of training and testing label will be divided by a global scale value which can then be used later to convert it back.

In [77]:
#normalising the values
normaliser_full = tf.keras.layers.Normalization(axis=-1)
normaliser_full.adapt(np.array(train_full_copy))

#dnn model
dnn_model_1 = tf.keras.Sequential([
      normaliser_full,
      tf.keras.layers.Dense(48, activation='relu'),
      tf.keras.layers.Dense(32, activation='relu'),
      tf.keras.layers.Dense(16, activation='relu'),
      tf.keras.layers.Dense(8, activation='relu'),
      tf.keras.layers.Dense(1)
  ])

dnn_model_1.compile(loss='mean_absolute_error',
                optimizer=tf.keras.optimizers.Adam(0.2),
                metrics = [tf.keras.metrics.RootMeanSquaredError(name='rmse')]
                )
     

Like the regression model, it requires multiple steps to fine tune and build a model. They are similar, however difference being with a neural network there can be many layers of nodes that all connect to each other passing data once it passes a threshold. In this particular model, it was decided to scale down the nodes the deeper into the network until the final output layer where there would only be one node. Like before using the same loss functions such as RMSE to calculate the errors within the model between the predicted and actual values.

In [80]:
#build model
history = dnn_model_1.fit(
    train_full_copy,
    training_labels_full,
    validation_split=0.2,
    verbose=1, 
    epochs=50)

Epoch 1/50
63/63 [==============================] - 0s 4ms/step - loss: 0.0124 - rmse: 0.0160 - val_loss: 0.0108 - val_rmse: 0.0141
Epoch 2/50
63/63 [==============================] - 0s 3ms/step - loss: 0.0145 - rmse: 0.0186 - val_loss: 0.0124 - val_rmse: 0.0166
Epoch 3/50
63/63 [==============================] - 0s 4ms/step - loss: 0.0126 - rmse: 0.0163 - val_loss: 0.0104 - val_rmse: 0.0141
Epoch 4/50
63/63 [==============================] - 0s 3ms/step - loss: 0.0165 - rmse: 0.0207 - val_loss: 0.0334 - val_rmse: 0.0359
Epoch 5/50
63/63 [==============================] - 0s 3ms/step - loss: 0.0175 - rmse: 0.0220 - val_loss: 0.0105 - val_rmse: 0.0144
Epoch 6/50
63/63 [==============================] - 0s 3ms/step - loss: 0.0135 - rmse: 0.0175 - val_loss: 0.0445 - val_rmse: 0.0466
Epoch 7/50
63/63 [==============================] - 0s 4ms/step - loss: 0.0194 - rmse: 0.0255 - val_loss: 0.0183 - val_rmse: 0.0208
Epoch 8/50
63/63 [==============================] - 0s 5ms/step - loss: 0.01

The summary of this model shows quite loss values however this may be due to the number of trips being scaled down. However with such low results it looks promosing. Again 80/20 split was decided for the validation method as it would allow the values to be checked, and in this instance there is a discrepency in some of the return results where the validated loss is almost double the training loss value. 

In [82]:
#predictions
dnn_model_1_results = dnn_model_1.evaluate(test_full_copy, testing_labels_full, verbose=1)
print(dnn_model_1_results)

new_data_dnn = np.array([12,1,51,0.00,52])
predictions = dnn_model_1.predict(new_data_dnn)*SCALE
print(predictions)

20/20 [==============================] - 0s 2ms/step - loss: 0.0226 - rmse: 0.0260
[0.022640321403741837, 0.026029072701931]
1/1 [==============================] - 0s 28ms/step
[[0.07946528]]


Loss results from the model are still low which should in theory mean that the accuracy of the dataset are good, however from the model's predictions it appears to still be inaccurate. The same inputs where passed into the model from the linear regression to determine if the neural network could be more accurate predictions. There could be amendments made to the dataset or the actual model itself perhaps using a fine tuning the optimiser or loss function.

In [84]:
#2014 experiment

#normalising the values
normaliser_full = tf.keras.layers.Normalization(axis=-1)
normaliser_full.adapt(np.array(train_2014_copy))

#dnn model
dnn_model_2 = tf.keras.Sequential([
      normaliser_full,
      tf.keras.layers.Dense(48, activation='relu'),
      tf.keras.layers.Dense(24, activation='relu'),
      tf.keras.layers.Dense(1)
  ])

dnn_model_2.compile(loss='mean_absolute_error',
                optimizer=tf.keras.optimizers.Adam(0.001))

Much like the previous model, it is made up of layers of nodes the difference being in this model is that there are less nodes and less layers. Furthermore the optimiser has been set to a lower rate of 0.001. 

In [85]:
#build model
history = dnn_model_2.fit(
    train_2014_copy,
    training_labels_2014,
    validation_split=0.15,
    verbose=1, 
    epochs=100)

Epoch 1/100
8/8 [==============================] - 1s 39ms/step - loss: 0.1937 - val_loss: 0.1125
Epoch 2/100
8/8 [==============================] - 0s 8ms/step - loss: 0.1261 - val_loss: 0.0881
Epoch 3/100
8/8 [==============================] - 0s 8ms/step - loss: 0.0793 - val_loss: 0.0732
Epoch 4/100
8/8 [==============================] - 0s 9ms/step - loss: 0.0652 - val_loss: 0.0676
Epoch 5/100
8/8 [==============================] - 0s 6ms/step - loss: 0.0553 - val_loss: 0.0660
Epoch 6/100
8/8 [==============================] - 0s 8ms/step - loss: 0.0481 - val_loss: 0.0567
Epoch 7/100
8/8 [==============================] - 0s 6ms/step - loss: 0.0409 - val_loss: 0.0547
Epoch 8/100
8/8 [==============================] - 0s 10ms/step - loss: 0.0354 - val_loss: 0.0493
Epoch 9/100
8/8 [==============================] - 0s 6ms/step - loss: 0.0313 - val_loss: 0.0471
Epoch 10/100
8/8 [==============================] - 0s 8ms/step - loss: 0.0271 - val_loss: 0.0461
Epoch 11/100
8/8 [=========

In this model there have been a few slight changes to the setup for example, since the data inputs is much smaller as it's using the 2014 dataset, the validation split was reduced to 0.15 which is 15%. The epochs where increased to 100 which will allow for the dataset to be passed through more. Again the loss results are relatively low however the validation loss is much higher which would suggest that the during the validation stage there where either incorrect estimates or the model won't make accurate predictions during testing.

In [86]:
#predictions
dnn_model_2_results = dnn_model_2.evaluate(test_2014_copy, testing_labels_2014, verbose=0)
print(dnn_model_2_results)

#new data
new_data_dnn_v2 = np.array([16,12,60,64,57,0.1,17,54,1027])
predictions = dnn_model_2.predict(new_data_dnn_v2) * SCALE
print(predictions)

0.027196988463401794
1/1 [==============================] - 0s 64ms/step
[[0.04488176]]


The model was passed in the same new dataset from the 2014 linear regression model to find out if it could make more accurate predictions. While the loss result is still relatively low the prediction value again is inaccurate. At this stage it may be that the data being used don't have a strong enough corelation to the number of collisions to make accurate enough predictions.

#Conclusion

In conclusion based on the results from both the neural network and the linear regression, it can be concluded there while there are definite trends between the weather and the number of collisions it doesn't appear to be enough to make accurate predictions. During the analysis stage there was slight corelations between weather features and the number of collisions however it's highly likely there are a number of other factors that could influence the number of collisions. It was observed especially in the year 2020 that there was noticeable spikes in the number of collision throughout the months until march, where it then drastically dropped due to covid which meant that the amount of people traveling had significantly dropped. Throughout the other years there where noticeable spikes during the holiday seasons again suggests traffic increases and could therefor affect the number of collisions. This combined with varying weather could infact effect the number of collisions however currently, based on the models it's apparent it may not be strong enough to make accurate enough predictions where emergency services are able to efficiently allocate resources. There could be further improvements to models, such as having more available data perhaps from bigger datasets, as well as further data cleansing on the original dataset which could therefor allow the model to make more accurate predictions. 


#Resources  
1. [Keras Loss Functions](https://neptune.ai/blog/keras-loss-functions)

#Reference List

- IBM. (2023). *What are neural networks?* [Online]. Available at: https://www.ibm.com/topics/neural-networks 
[Accessed on 4th May 2023]

- IBM. (2023). *Linear regression* [Online]. Available at: https://www.ibm.com/topics/linear-regression
[Accessed on 4th May 2023]

- Neptune Ai. (2023). *Keras Loss Functions* [Online]. Available at: https://neptune.ai/blog/keras-loss-functions
[Accessed on 4th May 2023]